internal/db/bootstrap_logs_test.go

```diff
@@ -0,0 +1,158 @@
+
+	var logs2 []dbmodel.BootstrapLog
+	err = s.Database.DB.Where("job_id = ?", jobId2).Order("line_number asc").Find(&logs2).Error
+	c.Assert(err, qt.IsNil)
+
+	c.Assert(logs2, qt.HasLen, 1)
+	c.Assert(logs2[0].LineNumber, qt.Equals, 1)
+	c.Assert(logs2[0].LogLine, qt.Equals, "Creating Juju controller \"diglett2\" on the-most-amazing-cloud")
+}
+
+func (s *dbSuite) TestBootstrapLogs_QueryBootstrapLogs(c *qt.C) {
+	ctx := c.Context()
+
+	err := s.Database.Migrate(ctx)
+	c.Assert(err, qt.IsNil)
+
+	// Query where the job doesn't exist
+	jobIdThatDoesntExist := uuid.New()
+	_, _, err = s.Database.QueryBootstrapLog(ctx, jobIdThatDoesntExist, 0)
+	c.Assert(err, qt.ErrorMatches, "job not found")
+
+	// Add job to reference
+	jobId, err := s.Database.AddJob(ctx, "test-job")
+	c.Assert(err, qt.IsNil)
+
+	// Query with no logs
+	loggies, nextOffsetVal, err := s.Database.QueryBootstrapLog(ctx, jobId, 0)
+	c.Assert(err, qt.IsNil)
+	c.Assert(loggies, qt.HasLen, 0)
+	c.Assert(nextOffsetVal, qt.Equals, 0)
+
+	// Test iterating through a simulated incoming logs
+	offsetTracker := 0
+	collectedLogs := make([]string, 0)
+
+	newLogs, nextOffsetValue, err := s.Database.QueryBootstrapLog(ctx, jobId, offsetTracker)
+	c.Assert(err, qt.IsNil)
+	offsetTracker = nextOffsetValue
+	collectedLogs = append(collectedLogs, newLogs...)
+
+	c.Assert(s.Database.AddBootstrapLog(ctx, jobId, "Creating Juju controller \"diglett\" on the-most-amazing-cloud"), qt.IsNil)
+	c.Assert(s.Database.AddBootstrapLog(ctx, jobId, "Fetching Juju agent binaries"), qt.IsNil)
+
+	newLogs, nextOffsetValue, err = s.Database.QueryBootstrapLog(ctx, jobId, offsetTracker)
+	c.Assert(err, qt.IsNil)
+	offsetTracker = nextOffsetValue
+	collectedLogs = append(collectedLogs, newLogs...)
+
+	c.Assert(s.Database.AddBootstrapLog(ctx, jobId, "Binaries contain gems"), qt.IsNil)
+	c.Assert(s.Database.AddBootstrapLog(ctx, jobId, "Gems appear to be very expensive"), qt.IsNil)
+
+	newLogs, nextOffsetValue, err = s.Database.QueryBootstrapLog(ctx, jobId, offsetTracker)
+	c.Assert(err, qt.IsNil)
+	offsetTracker = nextOffsetValue
+	collectedLogs = append(collectedLogs, newLogs...)
+
+	c.Assert(collectedLogs, qt.HasLen, 4)
+
+	newLogs, nextOffsetValue, err = s.Database.QueryBootstrapLog(ctx, jobId, offsetTracker)
+	c.Assert(err, qt.IsNil)
+	// This means no new logs have come in, but they may later, and the client should query again for logs
+	// after some time.
+	c.Assert(newLogs, qt.HasLen, 0)
+	c.Assert(nextOffsetValue, qt.Equals, offsetTracker)
+}
+
+// This test is a behaviour check, that is, we want our lock does indeed reject
+// on an ACCESS EXCLUSIVE mode when inserting new logs.
+func (s *dbSuite) TestBootstrapLogs_lockBootstrapLogs(c *qt.C) {
+	ctx := c.Context()
+
+	err := s.Database.Migrate(ctx)
+	c.Assert(err, qt.IsNil)
+
+	jobId, err := s.Database.AddJob(ctx, "test-job")
+	c.Assert(err, qt.IsNil)
+
+	finishTransaction := make(chan bool)
+	c.Cleanup(func() {
+		close(finishTransaction)
+	})
```

Comment:
Looks fine at a glance but just ensure this doesn't leak some resources like "tries to close DB but fails because the transaction was still busy"

Reply:
Yeah seemed to be ok when running test. Without, the routine is stuck, so I think it is ok.